File: backend/app/models/fact.py
from typing import List, Dict, Optional
from pydantic import BaseModel, Field
from datetime import datetime
import uuid

class FactVariants(BaseModel):
    standard: str = ""
    metric_focused: Optional[str] = None
    architecture_focused: Optional[str] = None
    leadership_focused: Optional[str] = None

class FactBlock(BaseModel):
    id: str = Field(default_factory=lambda: str(uuid.uuid4())[:8])
    company: str = "Unknown Company"
    role: str = "Software Engineer"
    date_range: str = "Present"
    project_context: str = ""
    category: str = "Backend / Architecture"
    sub_category: str = "General"
    
    # Primary polished text (Google X-Y-Z formula)
    refined_text: str
    
    # Pre-generated variants for different JD emphasis
    variants: FactVariants = Field(default_factory=FactVariants)
    
    # Tags & Entities
    tech_stack: List[str] = Field(default_factory=list)
    metrics: List[str] = Field(default_factory=list)
    
    # Control flags
    is_locked: bool = False
    is_verified: bool = False
    personal_notes: Optional[str] = None
    
    # Source traceability
    source_resume_ids: List[str] = Field(default_factory=list)
    raw_source_snippets: List[str] = Field(default_factory=list)
    
    created_at: str = Field(default_factory=lambda: datetime.now().isoformat())
    updated_at: str = Field(default_factory=lambda: datetime.now().isoformat())

class FactCreateRequest(BaseModel):
    company: str
    role: str
    date_range: str
    project_context: Optional[str] = ""
    category: str = "Backend / Architecture"
    sub_category: str = "General"
    refined_text: str
    tech_stack: List[str] = []
    metrics: List[str] = []
    personal_notes: Optional[str] = None

class FactUpdateRequest(BaseModel):
    company: Optional[str] = None
    role: Optional[str] = None
    date_range: Optional[str] = None
    project_context: Optional[str] = None
    category: Optional[str] = None
    sub_category: Optional[str] = None
    refined_text: Optional[str] = None
    tech_stack: Optional[List[str]] = None
    metrics: Optional[List[str]] = None
    is_locked: Optional[bool] = None
    is_verified: Optional[bool] = None
    personal_notes: Optional[str] = None

class FactPolishRequest(BaseModel):
    target_focus: str = "metric_focused"  # "metric_focused", "architecture_focused", "leadership_focused"
    user_instruction: Optional[str] = None

File: frontend/src/types/fact.ts
export interface FactVariants {
  standard: string;
  metric_focused?: string | null;
  architecture_focused?: string | null;
  leadership_focused?: string | null;
}

export interface FactBlock {
  id: string;
  company: string;
  role: string;
  date_range: string;
  project_context: string;
  category: string;
  sub_category: string;
  refined_text: string;
  variants: FactVariants;
  tech_stack: string[];
  metrics: string[];
  is_locked: boolean;
  is_verified: boolean;
  personal_notes?: string | null;
  source_resume_ids: string[];
  raw_source_snippets: string[];
  created_at: string;
  updated_at: string;
}

export interface ResumeMetadata {
  id: string;
  filename: string;
  file_type: string;
  file_size: number;
  uploaded_at: string;
  status: 'UPLOADED' | 'PARSED' | 'EXTRACTED' | 'ERROR';
  character_count: number;
  extracted_fact_count: number;
  error_message?: string | null;
}

export interface FactCreatePayload {
  company: string;
  role: string;
  date_range: string;
  project_context?: string;
  category: string;
  sub_category: string;
  refined_text: string;
  tech_stack: string[];
  metrics: string[];
  personal_notes?: string;
}

export interface FactUpdatePayload {
  company?: string;
  role?: string;
  date_range?: string;
  project_context?: string;
  category?: string;
  sub_category?: string;
  refined_text?: string;
  tech_stack?: string[];
  metrics?: string[];
  is_locked?: boolean;
  personal_notes?: string;
}

export interface AppStats {
  total_resumes: number;
  total_facts: number;
  locked_facts: number;
  unique_companies: number;
  unique_tech_count: number;
  categories_breakdown: Record<string, number>;
}
